### PR TITLE
Add World::trigger_ref and World::trigger_targets_ref

### LINKED ISSUE #14888: World::trigger variant that takes a &mut impl Event
## What problem does this solve or what need does it fill?

I want to perform cleanup after `Observer`s have been triggered with an event, given they are passed into observer systems as `&mut Event`, i.e.:

```rust
let mut foo: FooEvent = FooEvent { /* ... */ };
world.trigger_ref_mut(&mut foo);
// then do something with the modified `FooEvent`.
```

## What solution would you like?

```rust
impl World {
    pub fn trigger_ref_mut(&mut self, event: &mut impl Event);
    pub fn trigger_ref_mut(&mut self, event: &mut impl Event, targets: impl TriggerTargets);
}
```

## What alternative(s) have you considered?

Getting commands ordered in specific ways for applying.

## Additional context

I believe this ability is required for a possible "deferred egui" model that I'm ruminating on.

### PR DESCRIPTION
**Pull request by ItsDoot**
# Objective

Closes #14888.

## Solution

Add non-consuming trigger functions:

```rust
impl World {
    pub fn trigger_ref(&mut self, event: &mut impl Event);
    pub fn trigger_targets_ref(&mut self, event: &mut impl Event, targets: impl TriggerTargets);
}
```

## Testing

- Added two new tests, `observer_trigger_ref` and `observer_trigger_targets_ref`.

### CODE REVIEW

**Review by alice-i-cecile — COMMENTED** [review]
The same methods should be added to `Commands` for consistency (and the reasons listed in your motivation).

File: crates/bevy_ecs/src/observer/mod.rs [hunk 1]
```diff
@@ -311,16 +311,39 @@ impl World {
         self.spawn(Observer::new(system))
     }
 
-    /// Triggers the given `event`, which will run any observers watching for it.
+    /// Triggers the given [`Event`], which will run any [`Observer`]s watching for it.
+    ///
+    /// This method consumes the event, so it cannot be used after this method is called.
```
Comment (janhohenheim) [inline crates/bevy_ecs/src/observer/mod.rs:316]: My `Event`s usually implement `Copy`, so that's not really true

**Comment by ItsDoot** [comment]
> The same methods should be added to `Commands` for consistency (and the reasons listed in your motivation).

This isn't possible, as all `Command`s are required to be `'static`.

File: crates/bevy_ecs/src/observer/mod.rs (hunk 1 above)
Comment (ItsDoot) [reply crates/bevy_ecs/src/observer/mod.rs:316]:
```suggestion
    /// This method consumes the event, so it cannot be used after this method is called,
    /// unless the event implements `Copy`.
```

Would something like this work? My motivation for this PR involves non-`Copy` event types.

Comment (ItsDoot) [reply crates/bevy_ecs/src/observer/mod.rs:316]:
```suggestion
    /// While event types commonly implement [`Copy`], those that don't will be consumed and will no longer be accessible.
```

Alternatively

**Review by janhohenheim — CHANGES_REQUESTED** [review]
I'm a bit weary of giving `Event`s in observers special treatment when compared to `Event`s in `EventWriter`s.
Why is it not enough to let the observers change a component that you then query later?
If there is a reason, why does this not apply to `EventWriter::send` as well?

Also @alice-i-cecile I think adding this method to `Command`s would be misleading. If a user calls it on a command, they will read this description:

> this method is most useful when it's necessary to check
> or use the event after it has been modified by observers.

and probably think "ah, I see, the observer callbacks are all executed when I call `.trigger_ref`!" which is not the case at all. It makes more sense on `World`, as you can actually trigger the observers instantly there.

File: crates/bevy_ecs/src/observer/mod.rs (hunk 1 above)
Comment (janhohenheim) [reply crates/bevy_ecs/src/observer/mod.rs:316]: Both are fine for me :)

**Review by alice-i-cecile — APPROVED** [review]

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Actually, rescinding my approval. This is well documented and correctly implemented, but I'd like to see more justification / a proof of concept for why this is actually useful. Totally fine if that's external, but maybe experiment off of this branch and see if it solves your needs? There's a cost to introducing a bit of a weird API here, and without more motivation this is likely to get reverted in a cleanup in the future.

**Comment by janhohenheim** [comment]
(TIL the verb "to rescind")

**Comment by ItsDoot** [comment]
> I'd like to see more justification / a proof of concept for why this is actually useful. Totally fine if that's external, but maybe experiment off of this branch and see if it solves your needs? There's a cost to introducing a bit of a weird API here, and without more motivation this is likely to get reverted in a cleanup in the future.

Yep, that's exactly my plan. I should have a more concrete use case at some point soon. :+1:

File: crates/bevy_ecs/src/observer/mod.rs [hunk 2]
```diff
@@ -311,16 +311,41 @@ impl World {
...
+    /// If you need to use the event after triggering it, use [`World::trigger_ref`] instead.
     pub fn trigger(&mut self, event: impl Event) {
         TriggerEvent { event, targets: () }.trigger(self);
     }
 
-    /// Triggers the given `event` for the given `targets`, which will run any observers watching for it.
+    /// Triggers the given [`Event`] as a mutable reference, which will run any [`Observer`]s watching for it.
+    ///
+    /// Compared to [`World::trigger`], this method is most useful when it's necessary to check
+    /// or use the event after it has been modified by observers.
```
Comment (cBournhonesque) [inline crates/bevy_ecs/src/observer/mod.rs:326]: maybe "is most useful when you don't want to consume the event when triggering the observer"?

**Comment by aecsocket** [comment]
I have a concrete use-case for this in my networking library. When a client attempts to connect to a server in my crate, we trigger a `SessionRequest` event that contains connection info like address and headers. The consumer of the crate must then make a decision based on those headers on whether the client should be allowed to connect or not.

Without the ability to read back the triggered event, consumers must trigger a 2nd separate event `SessionResponse` to determine whether the client can connect or not. If they don't send this event, then the client never connects, and I would have to add more explicit handling to say, "if we don't get a SessionResponse by the end of this frame, disconnect them and log a warning".

I use a oneshot channel to forward the response to an async task. If I can't immediately send a value along this channel, then I have to keep this channel around in a component on the client's entity. This isn't a deal-breaker, but just harms ergonomics when there is a possible cleaner solution to my problem.

Current code:

```rs
// when we get a new connection request...
        let session = commands
            .spawn((
                ClientFrontend::Connecting {
                    // here's the oneshot channel over which the SessionResponse is sent
                    send_session_response: Some(connecting.send_session_response),
                    // ...
                },
            ))
            .set_parent(server)
            .id();
        let _ = connecting.send_session_entity.send(session);

        let request = SessionRequest {
            headers: connecting.headers, // ...
        };
        commands.trigger_targets(request, session);

// when the user responds to this request...
fn on_connection_response(
    trigger: Trigger<SessionResponse>,
    mut clients: Query<&mut ClientFrontend>,
) {
    let client = trigger.entity();
    let Ok(mut frontend) = clients.get_mut(client) else {
        return;
    };
    let ClientFrontend::Connecting {
        send_session_response,
        ..
    } = frontend.as_mut()
    else {
        return;
    };
    let Some(sender) = send_session_response.take() else {
        return;
    };

    let _ = sender.send(*trigger.event());
}
```

Ideally:

```rs

// when we get a new connection request...
        let session = commands
            .spawn(ClientFrontend::Connecting { .. })
            .set_parent(server)
            .id();
        let _ = connecting.send_session_entity.send(session);

        let mut request = SessionRequest {
            headers: connecting.headers, // ...
            response: SessionResponse::Undecided,
        };
        commands.add(move |world: &mut World| {
            world.trigger_targets_ref(&mut request, session);
            connecting.send_session_response.send(request.response);
        });
```

This also means that I can immediately validate whether the consumer of my crate actually set the SessionResponse, without having to wait until the next frame:
```rs
world.trigger_targets_ref(&mut request, session);

if request.response == SessionResponse::Undecided {
    warn!("User code did not provide a response to connecting client {session}");
}

connecting.send_session_response.send(request.response);
```

I expect that this approach of, "I'm sending out this event to ask if X can do Y", can be applied in a lot more situations as well.

> I'm a bit weary of giving Events in observers special treatment when compared to Events in EventWriters.
> Why is it not enough to let the observers change a component that you then query later?
> If there is a reason, why does this not apply to EventWriter::send as well?

IMO the fact that you have mutable world access and are immediately applying changes to the world marks a big difference between EventWriters and Triggers. Changing a component can be unergonomic (see my concrete use case above), and this pattern of "send an event -> allow consumers to change the outcome of the event -> read back the event" is only possible when you have this mutable world access. It's also an established pattern in other software e.g. [the Minecraft server software Paper](https://papermc.io/), where while listening to an event, you can choose to change certain parameters, or even cancel it entirely.

**Review by alice-i-cecile — APPROVED** [review]
Great, that was incredibly useful context. I'd love to see that make its way into an example at some point, but it's not needed for this PR.

@aecsocket, if you're happy with the code quality here, please leave an approval so we can get this merged in :)

File: crates/bevy_ecs/src/observer/mod.rs [hunk 3]
```diff
@@ -311,16 +311,41 @@ impl World {
...
+    pub fn trigger_ref(&mut self, event: &mut impl Event) {
+        TriggerEvent { event, targets: () }.trigger_ref(self);
+    }
+
+    /// Triggers the given [`Event`] for the given `targets`, which will run any [`Observer`]s watching for it.
+    ///
+    /// /// While event types commonly implement [`Copy`],
```
Comment (aecsocket) [inline crates/bevy_ecs/src/observer/mod.rs:333]: The double `///` is likely a copy-paste error?

**Review by pablo-lua — APPROVED** [review]
I think if would be better to have an actual example of how this can be useful in the trigger_ref docs, but just an small nit. I'm fine with it as it is

**Review by aecsocket — COMMENTED** [review]
LGTM apart from one comment

File: crates/bevy_ecs/src/observer/mod.rs (hunk 3 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/observer/mod.rs:333]:
```suggestion
    /// While event types commonly implement [`Copy`],
```

**Review by aecsocket — APPROVED** [review]

**Pull request merged**